garrysmod/lua/vgui/dnumslider.lua

```diff
@@ -10,7 +10,16 @@ function PANEL:Init()
 	self.TextArea:SetPaintBackground( false )
 	self.TextArea:SetWide( 45 )
 	self.TextArea:SetNumeric( true )
-	self.TextArea.OnChange = function( textarea, val ) self:SetValue( self.TextArea:GetText() ) end
+	self.TextArea:SetUpdateOnType( true )
+	self.TextArea.OnValueChange = function( textentry, val )

```

Comment:
This whole callback is unnecessary as the numslider's value is already clamped in its SetValue. Instead, just clamp the textentry's value in that function alongside the numslider so it isn't performed duplicitously.

Reply:
So do you advice just not to set slider's value in that function? Seems to work.
```lua
self.TextArea.OnValueChange = function( textentry, val )
	local cPos = self.TextArea:GetCaretPos()
	local clamped = math.Clamp( tonumber( val ) || 0, self:GetMin(), self:GetMax() )
	if clamped != val then
		self.TextArea:SetText( clamped )
		self.TextArea:SetCaretPos( cPos )
	end
end
```

Reply:
It works but it does the clamping twice, once in OnValueChange and another time in SetValue. You can optimise this by using the NumSlider's clamping from https://github.com/Facepunch/garrysmod/blob/057e56ed4d56fa6b177a44d3028473fbedcdc033/garrysmod/lua/vgui/dnumslider.lua#L111-L121 , ex.
```Lua
function PANEL:SetValue( val )

	val = math.Clamp( tonumber( val ) || 0, self:GetMin(), self:GetMax() )

	if ( self:GetValue() == val ) then return end

	self.Scratch:SetValue( val ) -- This will also call ValueChanged

	val = tostring( val )
	self.TextArea:SetText( val )
	self.TextArea:SetCaretPos( #val )

	self:ValueChanged( self:GetValue() ) -- In most cases this will cause double execution of OnValueChanged

end
```